tests/herbie.egg

```diff
@@ -0,0 +1,573 @@
+;; and a non-zero analysis.
+;; The non-zero analysis is built off the interval analysis (in order to prove
+;; that rewrites are sound, even if some parts of an expr can't be const-evaled)
+
+; TODO: unbounded intervals?
+(function hi (Math) Rational :merge (min old new))
+(function lo (Math) Rational :merge (max old new))
+(relation non-zero (Math))
+
+;; First, constant folding!
+;; We don't need an explicit constant folding analysis, we can just union
+;; with nums when we can
+
+; Cases
+(rewrite (Add (Num a) (Num b)) (Num (+ a b)))
+(rewrite (Sub (Num a) (Num b)) (Num (- a b)))
+(rewrite (Mul (Num a) (Num b)) (Num (* a b)))
+(rewrite (Div (Num a) denom) (Num (/ a b)) :when ((= denom (Num b)) (non-zero denom)))
+(rewrite (Pow (Num a) (Num b)) (Num res) :when ((= res (try-pow a b))))
+(rewrite (Neg (Num a)) (Num (neg a)))
+(rewrite (Sqrt (Num a)) (Num res) :when ((= res (try-sqrt a))))
+(rewrite (Cbrt (Num a)) (Num res) :when ((= res (try-cbrt a))))
+(rewrite (Fabs (Num a)) (Num (abs a)))
+(rewrite (Ceil (Num a)) (Num (ceil a)))
+(rewrite (Floor (Num a)) (Num (floor a)))
+(rewrite (Round (Num a)) (Num (round a)))
+(rewrite (Log (Num a)) (Num res) :when ((= res (try-log a))))
+
+;; To check if something is zero, we check that zero is not contained in the
+;; interval. There are two possible (overlapping!) cases:
+;; - There exists a lo interval, in which case it must be larger than 0
+;; - There exists a hi interval, in which case it must be smaller than 0
+;; This assumes that intervals are well-formed: lo <= hi at all times.
+(rule ((= l (lo e))
+       (> l (rational 0 1)))
+      ((non-zero e)))
+(rule ((= h (hi e))
+       (< h (rational 0 1)))
+      ((non-zero e)))
+
+(rule ((= e (Num ve)))
+      ((set (lo e) ve)
+       (set (hi e) ve)))
+
+;; The interval analyses are similar to the constant-folding analysis,
+;; except we have to take the lower/upper bound of the results we get
+(rule ((= e (Add a b))
+       (= la (lo a))
+       (= lb (lo b)))
+      ((set (lo e) (+ la lb))))
+(rule ((= e (Add a b))
+       (= ha (hi a))
+       (= hb (hi b)))
+      ((set (hi e) (+ ha hb))))
+      
+(rule ((= e (Sub a b))
+       (= la (lo a))
+       (= ha (hi a))
+       (= lb (lo b))
+       (= hb (hi b)))
+      ((set (lo e) 
+          (min (min (- la lb) (- la hb))
+               (min (- ha lb) (- ha hb))))
+       (set (hi e) 
+          (max (max (- la lb) (- la hb))
+               (max (- ha lb) (- ha hb))))))
+
+(rule ((= e (Mul a b))
+       (= la (lo a))
+       (= ha (hi a))
+       (= lb (lo b))
+       (= hb (hi b)))
+      ((set (lo e) 
+          (min (min (* la lb) (* la hb))
+               (min (* ha lb) (* ha hb))))
+       (set (hi e) 
+          (max (max (* la lb) (* la hb))
+               (max (* ha lb) (* ha hb))))))
+
+(rule ((= e (Div a b))
+       (= la (lo a))
+       (= ha (hi a))
+       (= lb (lo b))
+       (= hb (hi b))
+       (!= lb (rational 0 1))
+       (!= hb (rational 0 1)))
```

Comment:
I think these guards aren't needed since division is partial (even in the action).